### ISSUE TITLE
C++ shuffle multiplication kernel does not support operands with different truncation depths

### ISSUE THREAD

**Issue by inakleinbottle** [datasig-ac-uk/roughpy-jax#39]
The dense C++ shuffle-product kernel assumes that both operand buffers extend through the full output depth. This is incompatible with `st_mul` and `st_fma` operations whose operands have different truncation depths.
The Python operation layer correctly selects the appropriate result basis:

- `st_mul` selects the deeper operand basis.
- `st_fma` uses the accumulator basis.

It also passes the actual maximum degree of each operand to the C++ kernel. However, the kernel does not correctly respect these independent degree bounds.

### Root cause

Both implementations of `basic::st_fma` in:

```text
roughpy_compute/dense/basic/shuffle_tensor_product.hpp
```

iterate over output degrees using the left operand’s maximum degree:

```cpp
for (Degree out_deg = 1; out_deg <= lhs.max_degree(); ++out_deg)
```

They then access every decomposition of the output word without checking whether the corresponding degrees lie within the operand views:

```cpp
auto lhs_level = lhs.at_level(lhs_degree);
auto rhs_level = rhs.at_level(rhs_degree);
```

The v2 implementation has the equivalent problem for both the current and next levels.

This causes two failure modes:

1. When the left operand is shallower than the output, the kernel never computes the deeper output levels. They remain zero.
2. When the right operand is shallower, the kernel may access levels beyond the right operand’s buffer. Debug builds may trigger an assertion; release builds can perform an out-of-bounds read and produce undefined values, including `NaN`.

Because `st_mul` is implemented through the same shuffle FMA kernel, both `st_mul` and `st_fma` are affected.

### Minimal reproduction

```python
import jax.numpy as jnp
import roughpy_jax as rpj

shallow_basis = rpj.TensorBasis(2, 1)
deep_basis = rpj.TensorBasis(2, 2)

lhs = rpj.ShuffleTensor(
    jnp.arange(1, shallow_basis.size() + 1, dtype=jnp.float32),
    shallow_basis,
)
rhs = rpj.ShuffleTensor(
    jnp.arange(1, deep_basis.size() + 1, dtype=jnp.float32),
    deep_basis,
)

result = rpj.st_mul(lhs, rhs)
expected = rpj.st_mul(lhs.change_depth(2), rhs)

assert jnp.allclose(result.data, expected.data)
```

Before the temporary workaround, the mixed-depth result was:

```text
[1, 4, 6, 0, 0, 0, 0]
```

whereas the depth-aligned computation produced:

```text
[1, 4, 6, 12, 17, 18, 25]
```

Reversing the operand depths may result in invalid reads and nondeterministic output.

### Expected behaviour

The kernel should:

- iterate through the output view’s requested degree range;
- include only decompositions whose left and right degrees lie within the corresponding operand views;
- never access an operand level outside its declared minimum and maximum degrees;
- support physically shorter operand buffers when their declared maximum degrees are lower than the output depth.

### Temporary workaround
`DenseSTMul` and `DenseSTFma` currently re-depth their multiplicand arrays to the resolved output basis before calling the kernel. Their kernel degree arguments are also set to the output depth so the padded coefficients are treated as zero.

This avoids invalid accesses and restores correct results, but introduces unnecessary padding and potential device allocations.

### Proposed resolution

Update both v1 and v2 implementations of `basic::st_fma` to handle independently truncated operand views correctly. Once fixed, remove the operand re-depthing and degree normalization from:

```text
DenseSTMul
DenseSTFma
```

### Acceptance criteria

- Mixed-depth `st_mul` works in either operand order.
- Mixed-depth `st_fma` works when either multiplicand is shallower than the accumulator.
- No operand level is accessed outside its declared range.
- Existing same-depth behaviour remains unchanged.
- The temporary Python/JAX re-depthing workaround is removed.
- The mixed-depth tests in `test_st_mul.py` and `test_st_fma.py` pass without padding in the operation layer.
